### PR TITLE
Emit downlevel parameter initializers unless we are certain they don't have any side effects.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
i.e. if you have:

```function foo(a = "") { }```

Then we can emit that as:

```function foo(a) { }```

but if you have:

```function foo(a = bar()) { }```

then we need to emit that as:

```
function foo(a) {
    if (a === void 0) { a = bar(); }
}
```

--

This also unifies some of emitting code between single line function blocks and other types of functions.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -3930,15 +3932,117 @@ module ts {
...
+                    case SyntaxKind.BindingElement:
+                        return hasPossibleSideEffects((<BindingElement>node).name) ||
+                            hasPossibleSideEffects((<BindingElement>node).initializer);
+
+                    case SyntaxKind.Parameter:
+                        return hasPossibleSideEffects((<ParameterDeclaration>node).name) ||
+                            hasPossibleSideEffects((<ParameterDeclaration>node).initializer);
+
+                    case SyntaxKind.PropertyAccessExpression:
+                        return hasPossibleSideEffects((<PropertyAccessExpression>node).expression);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This should return `false` due to `get` accessors. Leave a comment about that.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -3930,15 +3932,117 @@ module ts {
...
+
+                    case SyntaxKind.ArrayLiteralExpression:
+                        return forEach((<ArrayLiteralExpression>node).elements, hasPossibleSideEffects);
+
+                    case SyntaxKind.ElementAccessExpression:
+                        return hasPossibleSideEffects((<ElementAccessExpression>node).expression) ||
+                            hasPossibleSideEffects((<ElementAccessExpression>node).argumentExpression);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I also believe this should return `false` because of `get` accessors.

File: src/compiler/emitter.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: Good point.

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**